File: pset6/credit/credit.py
from cs50 import get_string
import re

n = get_string("Number: ") # get the number as a string
size = len(n)

sum1 = 0

if size != 13 and size != 15 and size != 16: # verify if it is a card size
    print("INVALID")

# Luhn's algorithm
for i in range(size-2, -1, -2):
    cc = int(n[i]) * 2

    # if the product has two digits, divide them to sum
    if cc > 9:
        while cc > 0:
            digit = cc % 10
            sum1 += digit
            cc //= 10
    else:
        sum1 += cc

for i in range(size-1, -1, -2):
    sum1 += int(n[i])

if sum1 % 10 == 0:
    # check which card is
    if re.search('^3[47][0-9]{13}$', n):
        print("AMEX")
    elif re.search('^5[1-5][0-9]{14}$', n):
        print("MASTERCARD")
    elif re.search('^4[0-9]{12}([0-9]{3})?$', n):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

int main(void)
{
    string n = get_string("Numero: ");
    int sum = 0;
    int size = strlen(n);

    //convert ASCII values to real numbers
    for (int i = 0; i < size; i++)
    {
        n[i] -= '0';
    }

    //verify if it is a card size
    if (size != 13 && size != 15 && size != 16)
    {
        printf("INVALID\n");
    }
    else
    {
        //Luhn's algorithm
        for (int i = size - 2; i >= 0; i -= 2)
        {
            int v = n[i] * 2;

            //if the product has two digits, divide them to sum
            if (v > 9)
            {
                while (v > 0)
                {
                    int digit = v % 10;
                    sum += digit;
                    v /= 10;
                }
            }
            else
            {
                sum += v;
            }
        }

        for (int i = size - 1; i >= 0; i -= 2)
        {
            sum += n[i];
        }

        if (sum % 10 == 0)
        {
            //check which card is
            if (n[0] == 3 && (n[1] == 4 || n[1] == 7))
            {
                printf("AMEX\n");
            }
            else if (n[0] == 5 && (n[1] == 1 || n[1] == 2 || n[1] == 3 || n[1] == 4 || n[1] == 5))
            {
                printf("MASTERCARD\n");
            }
            else if (n[0] == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}